solc/CommandLineInterface.cpp

```diff
@@ -931,45 +932,61 @@ void CommandLineInterface::link()
 		// be just the cropped or '_'-padded library name, but this changed to
 		// the cropped hex representation of the hash of the library name.
 		// We support both ways of linking here.
-		librariesReplacements["__" + evmasm::LinkerObject::libraryPlaceholder(name) + "__"] = library.second;
-
-		string replacement = "__";
-		for (size_t i = 0; i < placeholderSize - 4; ++i)
-			replacement.push_back(i < name.size() ? name[i] : '_');
-		replacement += "__";
-		librariesReplacements[replacement] = library.second;
+		librariesReplacements[name] = library.second;
 	}
 
 	FileReader::StringMap sourceCodes = m_fileReader.sourceUnits();
+
 	for (auto& src: sourceCodes)
 	{
-		auto end = src.second.end();
-		for (auto it = src.second.begin(); it != end;)
+		// Map from how the libraries will be named inside the bytecode to their addresses.
+		map<size_t, string> linkReferences;
+		string bytecode;
+		vector<string> sourceLines;
+		boost::split(sourceLines, src.second, boost::is_any_of("\n"));
+
+		if (sourceLines.size() < 3)
+			// no bytecode found
+			continue;
+
+		size_t bytecodeIndex;
+		// Bytecode can be on line 0 or 3, depending on if file header is included
```

Comment:
Do we really want to account for the output formatting of the non-file-based commandline interface?

Reply:
Ideally yes, because this is required to allow the tests to pass although, I'm away that it's not the most elegant solution. I also considered using a regex to find the index instead of hard coding it but, this seemed simpler.

Reply:
I'd do one of these here:
1) Assume a rigid file structure and validate it more strictly. For example when you skip files shorter than 3 lines, verify that these lines do not contain the bytecode (i.e. a string consisting only of hex chars without whitespace and possibly placeholders between them)
2) Detect the bytecode and ignore everything else. This will be more robust against things like someone just taking the compiler output (with multiple bytecode pieces in it), putting it in a file and running the linker on it.

The current version is neither of these.

Given how the linker has been working so far (basically accepting anything, regardless of it's bytecode or not), I think (2) would be better.